crates/edr_provider/src/requests/eth/transactions.rs

```diff
@@ -1,6 +1,117 @@
-use edr_eth::{serde::ZeroXPrefixedBytes, transaction::EthTransactionRequest, B256};
+use edr_eth::{
+    remote,
+    serde::ZeroXPrefixedBytes,
+    transaction::{EthTransactionRequest, SignedTransaction},
+    SpecId, B256, U256,
+};
 
-use crate::{data::ProviderData, ProviderError};
+use crate::{
+    data::{BlockMetadataForTransaction, GetTransactionResult, ProviderData},
+    ProviderError,
+};
+
+const FIRST_HARDFORK_WITH_TRANSACTION_TYPE: SpecId = SpecId::BERLIN;
+
+pub async fn handle_get_transaction_by_hash(
+    data: &ProviderData,
+    transaction_hash: B256,
+) -> Result<Option<remote::eth::Transaction>, ProviderError> {
+    data.transaction_by_hash(&transaction_hash)
+        .await?
+        .map(get_transaction_result_to_rpc_result)
+        .transpose()
+}
+
+fn get_transaction_result_to_rpc_result(
+    result: GetTransactionResult,
+) -> Result<remote::eth::Transaction, ProviderError> {
+    fn gas_price_for_post_eip1559(
+        signed_transaction: &SignedTransaction,
+        block_metadata: Option<&BlockMetadataForTransaction>,
+    ) -> U256 {
+        let max_fee_per_gas = signed_transaction
+            .max_fee_per_gas()
+            .expect("Transaction must be post EIP-1559 transaction.");
+        let max_priority_fee_per_gas = signed_transaction
+            .max_priority_fee_per_gas()
+            .expect("Transaction must be post EIP-1559 transaction.");
+
+        if let Some(block_metadata) = block_metadata {
+            let base_fee_per_gas = block_metadata.base_fee_per_gas.expect(
+                "Transaction must have base fee per gas in block metadata if EIP-1559 is active.",
+            );
+            let priority_fee_per_gas =
+                max_priority_fee_per_gas.min(max_fee_per_gas - base_fee_per_gas);
+            base_fee_per_gas + priority_fee_per_gas
+        } else {
+            // We are following Hardhat's behavior of returning the max fee per gas for
+            // pending transactions.
+            max_fee_per_gas
+        }
+    }
+
+    let GetTransactionResult {
+        signed_transaction,
+        spec_id,
+        block_metadata,
+    } = result;
+
+    let gas_price = match &signed_transaction {
+        SignedTransaction::PreEip155Legacy(tx) => tx.gas_price,
+        SignedTransaction::PostEip155Legacy(tx) => tx.gas_price,
+        SignedTransaction::Eip2930(tx) => tx.gas_price,
+        SignedTransaction::Eip1559(_) | SignedTransaction::Eip4844(_) => {
+            gas_price_for_post_eip1559(&signed_transaction, block_metadata.as_ref())
+        }
+    };
+
+    let chain_id = match &signed_transaction {
+        // Following Hardhat in not returning `chain_id` for `PostEip155Legacy` legacy transactions
```

Comment:
@fvictorio, @alcuadrado is there an objection in us returning the `chain_id` here?

Reply:
To clarify, I place my assertion on the types here: https://github.com/NomicFoundation/hardhat/blob/06474681f72e1cd895abbec419f6f10be3d8e4ed/packages/hardhat-core/src/internal/hardhat-network/provider/output.ts#L70

Reply:
I think what Agost did is correct. Pre-EIP-155 transactions don't have the chainId anywhere, and post-EIP-155 have it encoded in the `v` value.

Since the JSON-RPC layers is not specified at all, I don't think it would be _wrong_ to include it, but I would err on the side of doing what Hardhat is doing for simplicity.